### PR TITLE
Fix PVC name at restore time

### PR DESCRIPTION
**Pull request by xing-yang**
This PR fixed a problem of PVC name at restore time to preserve the original PVC name in Guest Cluster.
Also refactored the code to make it more modular.

Tested on my Guest Cluster setup.

### CODE REVIEW

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/241/

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/242/

**Comment by xing-yang** [comment]
Pre-check failed because of this error:
```
14:52:52  Error response from daemon: login attempt to https://registry-1.docker.io/v2/ failed with status: 429 Too Many Requests
```

This PR only affects the Guest Cluster restore code path so it won't have impact on tests running on Vanilla cluster.

**Comment by lintongj** [comment]
> Pre-check failed because of this error:
> 
> ```
> 14:52:52  Error response from daemon: login attempt to https://registry-1.docker.io/v2/ failed with status: 429 Too Many Requests
> ```
> 
> This PR only affects the Guest Cluster restore code path so it won't have impact on tests running on Vanilla cluster.

The failure indicates only a rate-limiting error from dockerhub which CICD pipeline is doing `docker push`. No functional error was identified.

**Review by lintongj — COMMENTED** [review]
Looks good.

File: pkg/paravirt/paravirt_protected_entity_type_manager.go [hunk 1]
```diff
@@ -282,26 +252,140 @@ func (this *ParaVirtProtectedEntityTypeManager) CreateFromMetadata(ctx context.C
...
+		return nil, err
+	}
+
+	gcPVC, err := this.createGuestPVC(gcPVCNamespace, gcPVCName, gcPV.Name, svcPVC)
 	if err != nil {
-		return nil, errors.Wrapf(err, "Could not generate PV name in the Guest Cluster for Supervisor Cluster PVC %s/%s",
+		return nil, err
+	}
+
+	err = this.updateCloneFromSnapshotStatus(cloneFromSnapshotNamespace, cloneFromSnapshotName, svcClone)
```
Comment (lintongj) [inline pkg/paravirt/paravirt_protected_entity_type_manager.go:273]: minor comment here. what about the following style for the error checking? It is more compact and prevents the error from being spread outside the if block. 

```
if err := this.updateCloneFromSnapshotStatus(cloneFromSnapshotNamespace, cloneFromSnapshotName, svcClone); err != nil {
    ...
}

File: pkg/paravirt/paravirt_protected_entity_type_manager.go [hunk 2]
```diff
@@ -282,26 +252,140 @@ func (this *ParaVirtProtectedEntityTypeManager) CreateFromMetadata(ctx context.C
...
+	// Decode metadata, change namespace of PVC
+	// from Guest Cluster namespace to Supervisor Cluster namespace,
+	// and encode again before calling CreateFromMetadata
+	// NOTE(xyang): We need to make assumption that StorageClass is already created
+	// in Supervisor and Synchronized to Guest Cluster
+	// StorageClass in Guest must have StorageClass name of Supervisor
+	// parameters:
+	//   svStorageClass: gc-storage-profile
+	svcPVC := v1.PersistentVolumeClaim{}
+	err := svcPVC.Unmarshal(metadata)
```
Comment (lintongj) [inline pkg/paravirt/paravirt_protected_entity_type_manager.go:330]: same as above

File: pkg/paravirt/paravirt_protected_entity_type_manager.go [hunk 3]
```diff
@@ -313,65 +397,75 @@ func (this *ParaVirtProtectedEntityTypeManager) CreateFromMetadata(ctx context.C
...
-			this.logger.Infof("PVC %s/%s saved", pvc.Namespace, pvc.Name)
+			this.logger.Infof("PVC %s/%s saved", gcPVC.Namespace, gcPVC.Name)
 		}
 	}
-	this.logger.Infof("CreateFromMetadata: PVC %s/%s is created in the guest cluster", pvc.Namespace, pvc.Name)
+	this.logger.Infof("createGuestPVC: PVC %s/%s is created in the guest cluster", gcPVC.Namespace, gcPVC.Name)
 
 	// Wait for PVC and PV to bind to each other
-	err = astrolabe_pvc.WaitForPersistentVolumeClaimPhase(v1.ClaimBound, this.gcKubeClientSet, pvc.Namespace, pvc.Name, astrolabe_pvc.Poll, astrolabe_pvc.ClaimBindingTimeout, this.logger)
+	err := astrolabe_pvc.WaitForPersistentVolumeClaimPhase(v1.ClaimBound, this.gcKubeClientSet, gcPVC.Namespace, gcPVC.Name, astrolabe_pvc.Poll, astrolabe_pvc.ClaimBindingTimeout, this.logger)
```
Comment (lintongj) [inline pkg/paravirt/paravirt_protected_entity_type_manager.go:453]: same as above

File: pkg/paravirt/paravirt_protected_entity_type_manager.go [hunk 4]
```diff
@@ -381,47 +475,11 @@ func (this *ParaVirtProtectedEntityTypeManager) CreateFromMetadata(ctx context.C
 	clone.Status.ResourceHandle = svcClone.Status.ResourceHandle.DeepCopy()
 	_, err = this.gcBackupDriverClient.CloneFromSnapshots(cloneFromSnapshotNamespace).UpdateStatus(context.TODO(), clone, metav1.UpdateOptions{})
```
Comment (lintongj) [inline pkg/paravirt/paravirt_protected_entity_type_manager.go:476]: not part of your code in this commit, but the style of error checking mentioned can also be applied here.

**Pull request merged**